### PR TITLE
stop relying on reduction init statements

### PR DESCRIPTION
**Pull request by skimo-openhub**
Use dependence information instead.

### CODE REVIEW

File: tc/core/polyhedral/cuda/mapped_scop.cc [hunk 1]
```diff
@@ -230,12 +229,17 @@ bool MappedScop::detectReductions(detail::ScheduleTree* tree) {
   if (!isReductionMember(member, updates, scop())) {
     return false;
   }
-  // Order the init statements (if any) before the update statements
+  // Order the other statements (if any) before the update statements
```
Comment (ftynse) [inline tc/core/polyhedral/cuda/mapped_scop.cc]: Can't we have other statements that are not updates and should be ordered _after_ reduction updates?  Something like
```
O(i)  = 0 for i in 0:N     #order before
O(i) += A(i,j)
O(i)  = O(i) / M             #order after
```

with a schedule
```
band([{S1[i]->[i]; S2[i,j]->[i]; S3[i]->[i]}; {S1[i]->[0]; S2[i,j]->[j]; S3[i]->[M]}])
  sequence()
    filter(S1[i])
    filter(S2[i,j])
    filter(S3[i])
```

**Comment by skimo-openhub** [comment]
On Fri, May 25, 2018 at 05:47:59AM -0700, ftynse wrote:
> ftynse commented on this pull request.
> > @@ -230,12 +229,17 @@ bool MappedScop::detectReductions(detail::ScheduleTree* tree) {
>    if (!isReductionMember(member, updates, scop())) {
>      return false;
>    }
> -  // Order the init statements (if any) before the update statements
> +  // Order the other statements (if any) before the update statements
> 
> Can't we have other statements that are not updates and should be ordered _after_ reduction updates?  Something like

Sure. This commit doesn't pretend to handle all cases,
only that it handles more cases than before and (most
importantly) that it does not rely on explicit init statements.

skimo

**Comment by ftynse** [comment]
Looks good.  Can you open issues for "reorder statements before/after the reduction based on dependences" and "use statement ids instead of tensor names in halide reduction matching", and mention them in the code to keep track of our observations?

**Comment by ftynse** [comment]
addressed comments myself

**Comment by ftynse** [comment]
Ordering before/after is now #454 

And cleaning up tc2halide is #455.

**Pull request merged**